### PR TITLE
Backend API Testing

### LINKED ISSUE #147: Some sort of testing system
**Please describe what has to be done**
Should make sure all pages are working, regression testing, etc. Likely a ways down the road though

### LINKED ISSUE #345: Backend route testing, fixing: users: /:id/edit
**Please describe what has to be done**
Test the route for defined functionality, proper error handling, input validation, etc. **See comments for more details**. 

**Additional context**
Prerequisite: https://github.com/PollBuddy/PollBuddy/issues/344

### LINKED ISSUE #347: Backend route testing, fixing: users: /:id/
**Please describe what has to be done**
Test the route for defined functionality. 

**Additional context**
Prerequisite: https://github.com/PollBuddy/PollBuddy/issues/346

### LINKED ISSUE #549: /api/users/me fixes
**Describe the bug**
When calling this route, such as from the contact us page automatically, the backend will crash as there's no test to see if a user is actually logged in before trying to get their ID from the session. There's actually a todo item in the code for it too. 

Also, it should return the following fields: 
- FirstName
- FirstNameLocked
- LastName
- LastNameLocked
- UserName
- UserNameLocked
- Email
- EmailLocked
- SchoolAffiliation

### LINKED ISSUE #562: Backend route testing, fixing: users: /me/edit
**Please describe what has to be done**
- [x] Should accept a JSON Object
* Only send changed fields
  * Accepted fields:
    * FirstName: String
    * LastName: String
    * UserName: String
    * Email: String
    * Password: String
    * LogOutEverywhere: Boolean (required if password is included, send error otherwise) 
		
 - [x] Backend should check for the existence of each key and validate each, and then if no errors, overwrite each saved value with the provided ones. 
 - [x] Upon success, return result = success and data as the same that /:id would give so the frontend can update its cache. Upon failure, return result = failure and error message as whatever the validation error is, or that there was a database problem.
 - [x] Please also note that the fields above are written in PascalCase to match the database data. However, the backend is a translation layer for the frontend, and therefore it should accept and return data in camelCase instead, as that is the format the frontend uses. This should simply mean changing the case of the first letter.
 - [x] Permissions should be checked. Here, this means make sure the user is logged in. 
 - [x] Special cases: 
   - [x] Protected Values:
     - [x] Should respect values that have a `<keyname>Locked` set to `true` by not allowing them to be changed. This applies to `FirstName`, `LastName`, `UserName`, and `Email`. Additionally, if `SchoolAffiliation` is set, `Password` changing should be locked too. 
   - [x] Resetting password: If a password is provided, it should be hashed as in the registration functions, and if `LogOutEverywhere` is set, the function `logOutEverywhere(userid);` should be called. 

Edit 2021/11/09: Added LogOutEverywhere info

### LINKED ISSUE #570: Lock UserName in DB on signup
**Please describe what has to be done**
Currently, the `UserNameLocked` field is not being set during registration. Given that we currently do not allow username changes, this should be set to `true` upon registration. 

Edit: `false` to `true`

### PR DESCRIPTION
**Pull request by PrinxeShamar**
### Description of change(s), including why:
Updates to User Model including added new getUser, editUser, getUserGroups to do specific operations for the user this way taking the handling out of the router and simply calling the function in the routes. 

Updated the mongoConnection file to be able to handle an already initialized database. This way you can have a database you already created being accessed when you call mongoConnect.getDB().

Updated user router to user the newly created functions in the User Model.

Added Jest Tests for the user router. (See Wiki Page for more detail) Created a new express app with the user router, set up an in memory database and wrote tests to tests the different aspects of the user router and make sure everything is functioning.

Created new dockerfiles to run the new Jest Tests in docker.

### Type of change(s)

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Any work to still do or other concerns:

Standardized Status Codes
Fully Detailed Backend Testing Wiki
Tests for Polls and Groups router.

### Closing issues:

- Closes #570
- Closes #562
- Closes #549
- Closes #347
- Closes #345 
- Closes #147

### CODE REVIEW

File: PollBuddy-Server/backend/routes/users.js [hunk 1]
```diff
@@ -371,11 +374,11 @@ router.post("/register", function (req, res) {
           }
 
         } else {
-
+          
           // One result changed, therefore it worked.
           // Configure user data and save in session
           req.session.userData = {};
-          req.session.userData.userID = result.insertedId.str;
+          req.session.userData.userID = result.insertedId;
```
Comment (hbh7) [inline PollBuddy-Server/backend/routes/users.js:381]: As long as this works, let's keep it and other usages then don't have to convert to object ID from string, like they currently do.

File: docker-compose.test.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    build:
+      context: PollBuddy-Server/frontend
+      dockerfile: "Dockerfile.test"
+
+  backend:
+    build:
+      context: PollBuddy-Server/backend
+      dockerfile: "Dockerfile.test"
+
+volumes:
```
Comment (hbh7) [inline docker-compose.test.yml:13]: Delete volumes section

File: PollBuddy-Server/backend/routes/users.test.js [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+mockApp.use(express.urlencoded({ extended: false }));
+mockApp.use(function (req, res, next) {
+  req.session = session;
+  next();
+});
+mockApp.use("/api/users", usersRouter); 
+
+let app = supertest(mockApp);
+
+beforeAll(async () => {
```
Comment (hbh7) [inline PollBuddy-Server/backend/routes/users.test.js:45]: Indicate if this is run once or once per test

File: PollBuddy-Server/backend/routes/users.test.js [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+  process.env.DB_NAME = global.__MONGO_DB_NAME__;
+  process.env.JEST = "true";
+
+  const client = await MongoClient.connect(process.env.DB_URL);
+  const db = client.db(process.env.DB_NAME);
+  mongoConnection.setClient(client);
+  mongoConnection.setDB(db);
+});
+
+afterAll(() => {
```
Comment (hbh7) [inline PollBuddy-Server/backend/routes/users.test.js:56]: Same

File: PollBuddy-Server/backend/routes/users.test.js [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+  setTimeout(() => {
+    mongoConnection.getClient().close();
+  }, 1000);
+});
+
+afterEach(async () => {
```
Comment (hbh7) [inline PollBuddy-Server/backend/routes/users.test.js:62]: Switch to beforeEach

File: PollBuddy-Server/backend/routes/users.test.js [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+  }
+  
+  let user = createModel(userData, userSchema);
+  let res = await mongoConnection.getDB().collection("users").insertOne(user);
+  return res;
+};
+
+describe("/api/users/login", () => {
+
+  it("GET: fail", async () => {
```
Comment (hbh7) [inline PollBuddy-Server/backend/routes/users.test.js:93]: change fail to "route unavailable"

File: PollBuddy-Server/backend/routes/users.test.js [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+    await app.get("/api/users/login")
+      .expect(405)
+      .then((response) => {
+        expect(response.body.result).toBe("failure");
```
Comment (hbh7) [inline PollBuddy-Server/backend/routes/users.test.js:97]: Check for error message
Create issue to create globally defined / standardized error messages

File: PollBuddy-Server/backend/routes/users.test.js [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+        expect(response.body.data.lastName).toBe(testUser.LastName);
+        expect(response.body.data.userName).toBe(testUser.UserName);
+        expect(session).toHaveProperty("userData.userID", res.insertedId);  
+      });
+  });
+
+  it("POST: login with email success", async () => {
+    let res = await createUser();
+    let userLogin = {
+      userNameEmail: testUser.UserName,
```
Comment (hbh7) [inline PollBuddy-Server/backend/routes/users.test.js:123]: Change username to email

File: PollBuddy-Server/backend/routes/users.test.js [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+        expect(response.body.data.userName).toBe(testUser.UserName);
+
+        let res = await mongoConnection.getDB().collection("users").findOne({ 
+          UserName: testUser.UserName,
+          Email: testUser.Email,
+          FirstName: testUser.FirstName,
+          LastName: testUser.LastName
+        });
+
+        expect(res).toBeTruthy();
```
Comment (hbh7) [inline PollBuddy-Server/backend/routes/users.test.js:174]: Double check that res is false if findOne doesn't actually find anything

File: PollBuddy-Server/backend/routes/users.test.js [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+
+describe("/api/users/logout", () => {
+
+  it("GET: logout user success", async () => {
+    session = { userData: {} };
+    await app.get("/api/users/logout")
+      .expect(200)
+      .then(async (response) => {
+        expect(response.body.result).toBe("success");
+        expect(session).toEqual({});
```
Comment (hbh7) [inline PollBuddy-Server/backend/routes/users.test.js:189]: Check against just userData being deleted

File: PollBuddy-Server/backend/routes/users.test.js [hunk 11]
```diff
@@ -0,0 +1,545 @@
...
+
+describe("/api/users/me", () => {
+
+  it("GET: get logged in user", async () => {
+    let res = await createUser();
+    session = { userData: { userID: res.insertedId } };
+    await app.get("/api/users/me")
+      .expect(200)
+      .then(async (response) => {
+        expect(response.body.result).toBe("success");
```
Comment (hbh7) [inline PollBuddy-Server/backend/routes/users.test.js:211]: Also check for some returned data

File: PollBuddy-Server/backend/routes/users.test.js [hunk 12]
```diff
@@ -0,0 +1,545 @@
...
+      });    
+  });
+
+  it("GET: user not logged in", async () => {
+    await app.get("/api/users/me")
+      .expect(400)
+      .then(async (response) => {
+        expect(response.body.result).toBe("failure");
```
Comment (hbh7) [inline PollBuddy-Server/backend/routes/users.test.js:219]: Add todo to check against standard error messages

File: PollBuddy-Server/backend/routes/users.test.js [hunk 13]
```diff
@@ -0,0 +1,545 @@
...
+      });    
+  });
+
+  it("POST: fail", async () => {
+    await app.post("/api/users/me")
+      .expect(405)
+      .then((response) => {
+        expect(response.body.result).toBe("failure");
```
Comment (hbh7) [inline PollBuddy-Server/backend/routes/users.test.js:227]: Add todo to check against standard error messages

**Review by hbh7 — CHANGES_REQUESTED** [review]
- [ ] Add comments
- [x] Fill out PR (such as TODO for groups/poll tests, more tests, etc) 
- [ ] Jest/testing overview page in wiki (including terminology like what describe, it, etc. mean, how to write tests, how to run tests, etc.)
- [x] Docker - Don't copy in globalconfig

File: PollBuddy-Server/backend/routes/users.test.js (hunk 5 above)
Comment (PrinxeShamar) [reply PollBuddy-Server/backend/routes/users.test.js:62]: Done.

File: PollBuddy-Server/backend/routes/users.test.js (hunk 6 above)
Comment (PrinxeShamar) [reply PollBuddy-Server/backend/routes/users.test.js:93]: Done.

File: PollBuddy-Server/backend/routes/users.test.js (hunk 8 above)
Comment (PrinxeShamar) [reply PollBuddy-Server/backend/routes/users.test.js:123]: Done.

File: PollBuddy-Server/backend/routes/users.test.js (hunk 9 above)
Comment (PrinxeShamar) [reply PollBuddy-Server/backend/routes/users.test.js:174]: yea test fails if it doesn't find anything.

File: PollBuddy-Server/backend/routes/users.test.js (hunk 10 above)
Comment (PrinxeShamar) [reply PollBuddy-Server/backend/routes/users.test.js:189]: Done.

**Comment by PrinxeShamar** [comment]
Hey Hunter can you rereview this can make sure everything is good?

**Review by hbh7 — APPROVED** [review]
Looks good!

**Pull request merged**